File: leetcode/dynamic_programming/dungeon.cpp
class Solution {
public:
    int m,n;
    int calculateMinimumHP(vector<vector<int>>& dungeon) {
        //dp[i][j]表示从dp[i][j]到达dp[m-1][n-1]时的最低生命值
        m=dungeon.size(),n=dungeon[0].size();
        return dp(dungeon,0,0);
    }

    int dp(vector<vector<int>>& dungeon,int i,int j)
    {
        if(i>=m || j>=n){
            return INT_MAX;
        }
        if(i==m-1 && j==n-1){
            if(dungeon[i][j] >= 0){
                return 1;
            }
            return 1-dungeon[i][j];
        }
        int next = min(dp(dungeon,i,j+1),dp(dungeon,i+1,j));

        if(dungeon[i][j] >= next){
            return 1;
        }

        if(dungeon[i][j] < 0)
        {
            int minHp = 1-dungeon[i][j];
            return max(minHp,next);
        }
        else
        {
            return next - dungeon[i][j];
        }
    }
};

File: leetcode/dynamic_programming/dungeon.py
class Solution:
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
        m , n = len(dungeon) , len(dungeon[0])
        cache = [ [-1] * n for _ in range(m) ]
        def dp(x: int,y: int,dungeon) -> int:
            m , n = len(dungeon) , len(dungeon[0])
            if x >= m or y >= n : return 1000*400
            if x == m-1 and y == n-1:
                if dungeon[x][y] < 0 :return 1-dungeon[x][y]
                else : return 1
            if cache[x][y] != -1: return cache[x][y]
            opt = min(dp(x+1,y,dungeon) , dp(x,y+1,dungeon))
            res = 0
            if dungeon[x][y] >= opt : res = 1
            else: res = opt - dungeon[x][y]
            cache[x][y]=res
            return res
        return dp(0,0,dungeon)
          
